### PR TITLE
refactor(proto!): matches in recursive Document structure

### PR DESCRIPTION
**Pull request by hanxiao**
image

```
doc1 --matches-- doc3 (with score ref_id=1), doc4 (with score ref_id=1), doc7 (with score ref_id=1)
                    |
                    | --- matches -- doc6 (with score ref_id=3), doc9 (with score ref_id=3), doc10 (with score ref_id=3)
```

This design:

- improves the consistency of recursive structure introduced in #652
- allows advanced training/search paradigm working on a proximity graph/nearest-neighbor graph (which is quite common in ML). It opens a door to graph learning.

# Migration Guide:

- when looping the topk results, directly access the attribute `x` without `match.x`. i.e.
```python
for k in d.matches: 
  fn(k.match.id)
```
is now 
```python
for k in d.matches: 
  fn(k.id)
```

### CODE REVIEW

File: jina/proto/jina.proto [hunk 1]
```diff
@@ -109,6 +101,8 @@ message Document {
 
     // the embedding array of this doc
     NdArray embedding = 19;
+
+    Relevance score = 20;
```
Comment (JoanFM) [inline jina/proto/jina.proto:105]: should `score` and `matches`  be `either-or`? I guess a Score would be assigned to a `retrieved` document, and the `matches` would be added to the `query document`?

File: tests/unit/drivers/test_querylang_drivers.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        yield d
+
+
+class DummySegmenter(BaseSegmenter):
+
+    def craft(self, text, *args, **kwargs):
+        return [{'text': 'adasd' * j} for j in range(10)]
+
+
+class MyTestCase(JinaTestCase):
```
Comment (JoanFM) [inline tests/unit/drivers/test_querylang_drivers.py:28]: Could we rename these `MyTestCase` to something more descriptive? Like QueryLangTestCase? I am trying to do it slowly when I touch some part related to them

File: jina/proto/jina.proto (hunk 1 above)
Comment (hanxiao) [reply jina/proto/jina.proto:105]: setting `oneof` on `matches` and `score` will make the following pattern unimplementable.

```
doc1 --matches-- doc3 (with score ref_id=1), doc4 (with score ref_id=1), doc7 (with score ref_id=1)
                    |
                    | --- matches -- doc6 (with score ref_id=3), doc9 (with score ref_id=3), doc10 (with score ref_id=3)
```

Comment (JoanFM) [reply jina/proto/jina.proto:105]: Is it a pattern that can happen? or it is for the sake of full flexibility?

Comment (JoanFM) [reply jina/proto/jina.proto:105]: In theory, `doc1` should come from a `query set`, and `doc3, doc4 and doc7` from the `index set`. This bifurcation of `matches` to me would have a different `meaning`. How a `doc` from the `index set` has matches `with others`? Wouldn't that imply something like a kNN graph?

Comment (hanxiao) [reply jina/proto/jina.proto:105]: This design:
- improves the consistency of recursive structure introduced in #652 
- allows advanced training/search paradigm that requires a proximity graph/nearest-neighbor graph (which is quite common in ML). It opens the door to graph learning.

Comment (JoanFM) [reply jina/proto/jina.proto:105]: understood ;)

Comment (hanxiao) [reply jina/proto/jina.proto:105]: That being said, most of our examples can be implemented with just one level of recursion.

Comment (JoanFM) [reply jina/proto/jina.proto:105]: Yes, I see it more as a potential enabler for different features in the future.

**Comment by JoanFM** [comment]
The access to results feel more natural now!

**Pull request merged**